rules/cc_side_outputs.bzl

```diff
@@ -0,0 +1,307 @@
+            # Adding the rule name in front of the file name helps mitigate this
+            # issue.
+            "{}.{}.{}".format(target.label.name, source_file.basename, extension),
+        )
+        outputs.append(output_file)
+
+        #output_file_tmp = ctx.actions.declare_file(output_file.basename + ".tmp")
+
+        # C files are treated specially, and have different flags applied
+        # (e.g. --std=c11).
+        opts = ctx.fragments.cpp.copts + ctx.rule.attr.copts
+        if source_file.extension == "c":
+            opts += ctx.fragments.cpp.conlyopts
+        else:
+            opts += ctx.fragments.cpp.cxxopts + ctx.rule.attr.cxxopts
+
+        c_compile_variables = cc_common.create_compile_variables(
+            feature_configuration = feature_configuration,
+            cc_toolchain = cc_toolchain,
+            source_file = source_file.path,
+            user_compile_flags = opts + flags,
+            include_directories = depset(
+                direct = [src.dirname for src in cc_compile_ctx.headers.to_list()],
+                transitive = [cc_compile_ctx.includes],
+            ),
+            quote_include_directories = cc_compile_ctx.quote_includes,
+            system_include_directories = cc_compile_ctx.system_includes,
+            framework_include_directories = cc_compile_ctx.framework_includes,
+            preprocessor_defines = depset(
+                direct = ctx.rule.attr.local_defines,
+                transitive = [cc_compile_ctx.defines],
+            ),
+        )
+
+        command_line = cc_common.get_memory_inefficient_command_line(
+            feature_configuration = feature_configuration,
+            action_name = ACTION_NAMES.c_compile,
+            variables = c_compile_variables,
+        )
+        env = cc_common.get_environment_variables(
+            feature_configuration = feature_configuration,
+            action_name = ACTION_NAMES.c_compile,
+            variables = c_compile_variables,
+        )
+
+        ctx.actions.run_shell(
+            tools = [
+                ctx.file._cleanup_script,
+            ],
+            arguments = [
+                c_compiler_path,
+                ctx.file._cleanup_script.path,
+                output_file.path,
+            ] + command_line,
+            env = env,
+            inputs = depset(
+                [source_file],
+                transitive = [
+                    cc_toolchain.all_files,
+                    cc_compile_ctx.headers,
+                ],
+            ),
+            outputs = [output_file],
+            command = """
+                CC=$1; shift
+                UNTAB=$1; shift
+                OUT=$1; shift
+                $CC -o - $@ | $UNTAB > $OUT
+            """,
+        )
+
+    return [CcSideProductInfo(files = depset(
+        direct = outputs,
+        transitive = transitive,
+    ))]
+
+def _cc_assembly_aspect_impl(target, ctx):
+    return _cc_compile_different_output(target, ctx, "s", ["-S"])
```

Comment:
TODO:
Emit assembly with original source as comments.
Emit other compiler outputs (e.g. preprocessed source).

Ok to do this in a follow-up PR.